### ISSUE TITLE
Fail to install dependencies on Windows

### ISSUE THREAD

**Issue by nvdaes** [edrlab/thorium-reader#1132]
tested locally and via GitHub Actions, where system requirements seems to be satisfied:

https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md

Workflow file:

https://github.com/nvdaes/readium-desktop/blob/actions/.github/workflows/build.yaml

Results:

https://github.com/nvdaes/readium-desktop/runs/884714314

This doesn't make possible test changes for possible pull requests.

**Comment by danielweck** [comment]
I think you need `runs-on: windows-2016` ( 
https://github.com/actions/virtual-environments/blob/master/images/win/Windows2016-Readme.md
 ) instead of `runs-on: windows-latest` ( https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md ):

https://github.com/nvdaes/readium-desktop/blob/38ac5dca33353b4ca5423a885fd2c4e7cb937e51/.github/workflows/build.yaml#L8

This is because some of Thorium's native dependencies compile with an older version of Visual Studio / MSBUILD component. I am seeing the following line in your GitHub Actions log, which indicates a more recent component which I think is not yet supported by Thorium's native dependencies:

`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\15.0\Bin\MSBuild.exe`

**Comment by danielweck** [comment]
To be clear: instead of Visual Studio 2019, I believe Thorium's native dependencies require Visual Studio 2017 (not 2015).

**Comment by danielweck** [comment]
Another idea: instead of `runs-on: windows-latest`, try `runs-on: vs2017-win2016`.

**Comment by danielweck** [comment]
Note that Thorium's automated CI builds are currently managed by AppVeyor (for Windows), and as you can see we enforce Visual Studio 2017 in order to instantiate the correct required version of MSBuild:

https://github.com/edrlab/thorium-reader/blob/952d899955d7e4450ef63f9286bff07fc91081b2/appveyor.yml#L1

MacOS and Linux CI is provided by Travis, which Thorium also used for Windows builds but we switched to AppVeyor due to compiler issues:

https://github.com/edrlab/thorium-reader/blob/952d899955d7e4450ef63f9286bff07fc91081b2/.travis.yml#L45

**Comment by nvdaes** [comment]
OK, I will try to use Windows 2016 and VS 2017 with GitHub Actions and
will come back with results. My computer has small free space and I
can have problems due to this fact, so sometimes it's better for me to
use GitHub Actions, which is simpler than appveyor for me.

2020-07-23 10:17 GMT+02:00, Daniel Weck <notifications@github.com>:

**Comment by danielweck** [comment]
Note that the `microsoft/setup-msbuild` action with `vs-version: '15.4'` would not work on `windows-latest` / `windows-2019`, because MSBuild only ships modern versions in this environment.

```
- name: Add msbuild to PATH
  uses: microsoft/setup-msbuild@v1.0.1
    with:
      vs-version: '15.4'
```

https://github.com/microsoft/setup-msbuild

**Comment by danielweck** [comment]
Note that the deprecation notice of `windows-2016` has been corrected, to extend the lifetime of this GitHub Actions environment:

https://github.com/actions/virtual-environments/issues/68#issuecomment-602652021

**Comment by danielweck** [comment]
@nvdaes correction in your GitHub Actions configuration:
`npm run package` should be `npm run package:win`
https://github.com/nvdaes/readium-desktop/blob/38ac5dca33353b4ca5423a885fd2c4e7cb937e51/.github/workflows/build.yaml#L22

**Comment by danielweck** [comment]
Interestingly, `windows-latest` / `windows-2019` includes the `Microsoft.VisualStudio.Component.VC.v141.x86.x64` component which should (if I understand correctly) provide support for Visual Studio 2017 (`VC.141` => `msvc2017`).

https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md#workloads-components-and-extensions

So perhaps the official documentation is out of sync with the actual GitHub Action environment, or maybe an additional command line call is needed to switch to the older version of MSBuild. I found this example which apparently adds 14 minutes to the setup process (this is to add XP libs):

```
- name: Add V7.1A SDK
      run: |
        "C:\Program Files (x86)\Microsoft Visual Studio\Installer\vs_installer.exe" modify --channelId VisualStudio.16.Release --productId Microsoft.VisualStudio.Product.Enterprise -q --add Microsoft.VisualStudio.Component.WinXP
      shell: cmd
```

**Comment by nvdaes** [comment]
OK, I will test in the evening and, if GitHub Actions are performed,
we can close this issue since probably the problem is in my computer.
Then if I want to send a PR I will create an artifact with GitHub
Actions and will install it locally for testing.

2020-07-23 10:49 GMT+02:00, Daniel Weck <notifications@github.com>:

**Comment by nvdaes** [comment]
Hello. These are results according to my tests:

- Dependencies can be installed using windows-16 (With Visual Studio
2017) in GitHub Actions.
- run: npm run __postinstall is not recognized.
- Without this command, run: npm start is cancelled after more than 6 hours:

https://github.com/nvdaes/readium-desktop/runs/905168758

BTW:
- npm run package:win is not documented in readme.

2020-07-23 11:10 GMT+02:00, Noelia <nrm1977@gmail.com>:

**Comment by danielweck** [comment]
Sorry, `npm run __postinstall` is not necessary anymore, as the included `postinstall` NPM script is automatically invoked after `npm install` or `npm ci`.

**Comment by danielweck** [comment]
`npm run start` and `npm run start:dev` launch the app and block the process until `CTRL+C` (or other SIGNAL interrupt), so this should not be invoked in CI environments.

**Comment by nvdaes** [comment]
Thanks for your two last comments. Perhaps readme.md maybe updated removing npm run postinstall (with underscores) and where the artifacts are located, and mentioning package:win. I will try again.
Also, should be the yaml file be included as a link in the wiki or something, including jobs for MacOS and Linux, in case someone cannot use his or her computer to build executables for testing?

Enviado desde mi iPhone

**Comment by nvdaes** [comment]
I think we can close this issue, since dependencies can be installed.
I have build a GitHub workflow in GitHub Actions and now the exe file
is properly built and I can download the artifact:

Workflow:
https://github.com/nvdaes/readium-desktop/actions/runs/181431129

Artifact with exe file (zipped):
https://github.com/nvdaes/readium-desktop/suites/960914224/artifacts/12082953

Feel free to close this or I may do it.


2020-07-24 15:42 GMT+02:00, nvdaes <notifications@github.com>:

**Comment by danielweck** [comment]
Excellent work, well done!

Timely work too, as I have recently been working on a build-related issue, and I have been considering for some time now migrating from AppVeyor (Windows) and Travis (Linux + MacOS) to GitHub Actions. I will report back here with my findings.

Closing the issue.

**Comment by nvdaes** [comment]
OK, I will try to build the Windows exe with latest-windows, may not
be possible. I don't have Linux nor MacOS, but I will be happy hearing
your findings. :)

2020-07-24 19:50 GMT+02:00, Daniel Weck <notifications@github.com>:

**Comment by danielweck** [comment]
Hello, I have now deprecated Travis and AppVeyor Continuous Integration services. Thorium now builds and deploys artefacts to the GitHub releases for Windows, Mac and Linux via GitHub Actions workflow:
https://github.com/edrlab/thorium-reader/blob/develop/.github/workflows/main.yml

Note that most of the hard work to delete + update the CI tagged-release stream, is actually implemented by this NodeJS script which uses the GitHub REST API (the off-the-shelf GitHub Actions are commented out in the YAML configuration, as every other method I tested had major pitfalls):
https://github.com/edrlab/thorium-reader/blob/develop/scripts/release-github.js

In other words, we do not use the official GitHub Actions "artefacts" feature, as this would quickly eat-up the bandwidth/storage allowance of the free service (apparently there is no easy way to delete temporary build artefacts, so instead Thorium continues to upload directly to GitHub releases, as with Travis and AppVeyor before).
